volatility/framework/plugins/windows/dumpfiles.py

```diff
@@ -0,0 +1,245 @@
+            requirements.PluginRequirement(name='handles', plugin=handles.Handles, version=(1, 0, 0))
+        ]
+
+    def dump_file_producer(self, file_object: interfaces.objects.ObjectInterface,
+                           memory_object: interfaces.objects.ObjectInterface,
+                           layer: interfaces.layers.DataLayerInterface,
+                           desired_file_name: str) -> str:
+        """Produce a file from the memory object's get_available_pages() interface.
+
+        :param file_object: the parent _FILE_OBJECT
+        :param memory_object: the _CONTROL_AREA or _SHARED_CACHE_MAP
+        :param layer: the memory layer to read from
+        :param desired_file_name: name of the output file
+        :return: result status
+        """
+        filedata = interfaces.plugins.FileInterface(desired_file_name)
+        try:
+            # Description of these variables:
+            #   memoffset: offset in the specified layer where the page begins
+            #   fileoffset: write to this offset in the destination file
+            #   datasize: size of the page
+            for memoffset, fileoffset, datasize in memory_object.get_available_pages():
+                data = layer.read(memoffset, datasize, pad = True)
+                filedata.data.seek(fileoffset)
+                filedata.data.write(data)
+
+            # Avoid writing files to disk if they are going to be empty or all zeros.
+            cached_length = len(filedata.data.getvalue())
+            if cached_length == 0 or filedata.data.getvalue().count(0) == cached_length:
+                result_text = "No data is cached for the file at {0:#x}".format(file_object.vol.offset)
+            else:
+                self.produce_file(filedata)
+                result_text = "Stored {}".format(filedata.preferred_filename)
+        except exceptions.InvalidAddressException:
+            result_text = "Unable to dump file at {0:#x}".format(file_object.vol.offset)
+
+        return result_text
+
+    def process_file_object(self, file_obj: interfaces.objects.ObjectInterface) -> Tuple:
+        """Given a FILE_OBJECT, dump data to separate files for each of the three file caches.
+
+        :param file_object: the FILE_OBJECT
+        """
+
+        # Filtering by these types of devices prevents us from processing other types of devices that
+        # use the "File" object type, such as \Device\Tcp and \Device\NamedPipe.
+        if file_obj.DeviceObject.DeviceType not in [FILE_DEVICE_DISK, FILE_DEVICE_NETWORK_FILE_SYSTEM]:
+            vollog.log(constants.LOGLEVEL_VVV,
+                       "The file object at {0:#x} is not a file on disk".format(file_obj.vol.offset))
+            return
+
+        # Depending on the type of object (DataSection, ImageSection, SharedCacheMap) we may need to
+        # read from the memory layer or the primary layer.
+        memory_layer = self.context.layers["memory_layer"]
```

Comment:
This is assumes that every plugin will always create a primary layer that manages to construct a memory layer, and that it's called memory layer.  This breaks if there's more than one intel layer floating around.  This either needs to check that primary is an intel layer, and then can ask it for its config option `base_layer`, or you can try to traverse `layer.dependencies` but that will list every layer it sits on (including any swap layers).

Reply:
Not `base_layer`, `memory_layer`, sorry.  5:S